### DISCUSSION TITLE
[4.x] Fs, volumes during testing

### DISCUSSION THREAD

**Discussion by qrazi** [craftcms/cms#14005] [Q&A] [answered]
Background: I'm finally getting round to upgrading a rather large and customized multisite Craft install to 4.x. We have, also devs, our volumes all configured to S3 buckets, since there are several GBs of assets that we want to use also in dev and staging environments.

For running Codeception tests in in 3.x, I "abuse" our AssetFixture class's `init` method to manually update all volumes from S3 to Local and set the path settings local as well. This is obvious a necessary thing, as you'd want tests to run in an isolated environment, that can be build and torn down entirely from code - I would not want to require a connection to S3, for performance and cost reasons (cost lesser of the two).

Now my question; how can I achieve this in Craft 4? How can I isolate the location of assets for test runs?

**Comment by brandonkelly** [comment, marked as answer]
In Craft 4, you can create separate filesystems for each environment: a Local filesystem for local development, and an S3 filesystem for production. Then, create a new environment variable in `.env` that is set to the Local filesystem’s handle, e.g.

```sh
IMAGES_FS=localImages
```

Then in your volume’s settings, set its filesystem to that same environment variable name (e.g. `$IMAGES_FS`). It will then use whatever filesystem the environment variable is referring to. Locally that could be `localImages`, but on production you could set the same environment variable to `s3Images` instead.

**Reply by qrazi** [reply]
Thanks a million!
